### PR TITLE
Show number of results in requirement and policy views

### PR DESCRIPTION
**Pull request by tadhg-ohiggins**
There are things to count. /
Count them, and birds above, /
And beams of sunlight.

Addresses #236.

### CODE REVIEW

File: ui/__tests__/components/thing-counters.test.jsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import ThingCounter from '../../components/thing-counters';
+import mockRouter from '../util/mockRouter';
+
+
+describe('<ThingCounter />', () => {
+  it('states zero results when there are no results', () => {
+    const context = { router: mockRouter() };
+    const result = shallow(<ThingCounter count={0} chunk={0} singular={ "requirement" } plural={ "requirements" }/>, { context });
+    expect(result.text()).toMatch(/No requirements match your search, try removing some filters to see more results./);
+    expect(result.find("div").length).toEqual(1);
```
Comment (cmc333333) [inline ui/__tests__/components/thing-counters.test.jsx]: NBD, but there's a `expect(...).toHaveLength()` function, if you like that more.

File: ui/__tests__/components/thing-counters.test.jsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import React from 'react';
+
+import ThingCounter from '../../components/thing-counters';
+import mockRouter from '../util/mockRouter';
+
+
+describe('<ThingCounter />', () => {
+  it('states zero results when there are no results', () => {
+    const context = { router: mockRouter() };
+    const result = shallow(<ThingCounter count={0} chunk={0} singular={ "requirement" } plural={ "requirements" }/>, { context });
```
Comment (cmc333333) [inline ui/__tests__/components/thing-counters.test.jsx]: I _think_ you can just do `singular="requirement"` here. You may also be able to do `count=0` rather than using so many curlies.

File: ui/__tests__/components/thing-counters.test.jsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    const context = { router: mockRouter({ query: { topic: 'something' } }) };
+    const result = shallow(<ThingCounter count={1} chunk={0} singular={ "requirement" } plural={ "requirements" }/>, { context });
+    expect(result.text()).toMatch(/1 requirement matches your search./);
+    expect(result.find("div").length).toEqual(1);
+    expect(result.find(".alert .p1 .m1 .border").length).toEqual(0);
+  });
+  it('states two results when there are two filtered results', () => {
+    const context = { router: mockRouter({ query: { topic: 'something' } }) };
+    const result = shallow(<ThingCounter count={2} chunk={0} singular={ "requirement" } plural={ "requirements" }/>, { context });
+    expect(result.text()).toMatch(/2 requirements match your search./);
```
Comment (cmc333333) [inline ui/__tests__/components/thing-counters.test.jsx]: 2 requirements, ah ah.

File: ui/__tests__/components/thing-counters.test.jsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    expect(result.find("div").length).toEqual(1);
+    expect(result.find(".alert .p1 .m1 .border").length).toEqual(0);
+  });
+  it('states one total result when there is one total result', () => {
+    const context = { router: mockRouter() };
+    const result = shallow(<ThingCounter count={1} chunk={0} singular={ "requirement" } plural={ "requirements" }/>, { context });
+    expect(result.text()).toMatch(/1 requirement./);
```
Comment (cmc333333) [inline ui/__tests__/components/thing-counters.test.jsx]: I may be missing something, but is this test different than 'states one result when there is one filtered result'? I think the period here matches any character.

Maybe you wanted a mock router with query params?

File: ui/__tests__/components/thing-counters.test.jsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    expect(result.find("div").length).toEqual(1);
+    expect(result.find(".alert .p1 .m1 .border").length).toEqual(0);
+  });
+  it('states two total results when there are two two results', () => {
```
Comment (cmc333333) [inline ui/__tests__/components/thing-counters.test.jsx]: Did you want "two total results" here?

Same question re: how this differs from 'states two results when there are two filtered results'

File: ui/components/policies/policies-view.jsx [hunk 6]
```diff
@@ -2,10 +2,16 @@ import React from 'react';
 
 import Pagers from '../pagers';
 import Policy from './policy-view';
+import ThingCounter from '../thing-counters';
 
 export default function PoliciesView({ policies, count, topicsIds }) {
+  const chunk = 25;
```
Comment (cmc333333) [inline ui/components/policies/policies-view.jsx]: It looks like the chunk size isn't used in `ThingCounter` -- can we remove it throughout?

File: ui/components/thing-counters.jsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+
+export default function ThingCounter({ count, singular, plural }, { router }) {
+  const query = router.location.query;
+  const queryKeys = Object.keys(query);
+  const noun = count === 1 ? singular : plural;
+  const verb = count === 1 ? 'matches' : 'match';
+  const classes = count === 0 ? 'alert p1 m1 border' : '';
+
+  const unfiltered = queryKeys.length === 0 || (queryKeys.length === 1 && queryKeys.indexOf('page') !== -1);
```
Comment (cmc333333) [inline ui/components/thing-counters.jsx]: This is a good, descriptive variable name. The logic feels a little hard to follow, so I'll throw out some odds and ends that may help in case you aren't familiar:
* `Object.assign({}, router.location.query)` would make a shallow copy of the query
* `del someObject.key` follows Python expectations for deleting an attribute. It's a noop if `key` isn't present
* `someArray.includes(key)` returns a bool and should be polyfilled for browser support already

Maybe we do something like:
```js
const queryWithoutPage = Object.assign({}, router.location.query);
del queryWithoutPage.page

const unfiltered = Object.keys(queryWithoutPage).length > 0
```
That doesn't feel particularly elegant, either...

File: ui/components/thing-counters.jsx [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  const searchText = unfiltered ? '.' : ` ${verb} your search.`;
+
+  const noMatches = `No ${noun} ${verb} your search, try removing some filters to see more results.`;
+  const someMatches = `${count} ${noun}${searchText}`;
+
+  const text = count === 0 ? `${noMatches}` : `${someMatches}`;
```
Comment (cmc333333) [inline ui/components/thing-counters.jsx]: Since we're taking the whole string, we can simplify this to:
```js
const text = count === 0 ? noMatches : someMatches;
```

File: ui/components/thing-counters.jsx [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  return (
+    <div className={classes}>
+      {`${text}`}
```
Comment (cmc333333) [inline ui/components/thing-counters.jsx]: Similarly, this can be `{text}`

File: ui/__tests__/components/thing-counters.test.jsx (hunk 5 above)
Comment (tadhg-ohiggins) [inline ui/__tests__/components/thing-counters.test.jsx]: Yup, I meant “two total results”.

When the result isn't filtered, i.e. when there are no query parameters, then there's no “that match your search” rider to the sentence.

File: ui/components/policies/policies-view.jsx (hunk 6 above)
Comment (tadhg-ohiggins) [inline ui/components/policies/policies-view.jsx]: I missed that one, and yes, it can go.

File: ui/components/thing-counters.jsx [hunk 10]
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+
+export default function ThingCounter({ count, singular, plural }, { router }) {
+  const query = router.location.query;
+  const queryKeys = Object.keys(query);
+  const noun = count === 1 ? singular : plural;
+  const verb = count === 1 ? 'matches' : 'match';
+  const classes = count === 0 ? 'alert p1 m1 border' : '';
```
Comment (cmc333333) [inline ui/components/thing-counters.jsx]: I think we'll need to include the associated SASS edits, too: https://github.com/18F/omb-eregs/pull/416/files#diff-a17de2846fcbead1087a20fc2730d26d

File: ui/components/thing-counters.jsx (hunk 8 above)
Comment (tadhg-ohiggins) [inline ui/components/thing-counters.jsx]: Probably! I find the JSX syntax somewhat awkward, as you can likely tell…

File: ui/__tests__/components/thing-counters.test.jsx (hunk 5 above)
Comment (cmc333333) [inline ui/__tests__/components/thing-counters.test.jsx]: Aha, I mixed up the two different types of test. The first two cover the extra query string, my bad.

File: ui/__tests__/components/thing-counters.test.jsx (hunk 4 above)
Comment (cmc333333) [inline ui/__tests__/components/thing-counters.test.jsx]: Apologies here as well. The period might still be capturing the wrong thing, but I mixed up the crux of the issue.

File: ui/components/thing-counters.jsx (hunk 7 above)
Comment (tadhg-ohiggins) [inline ui/components/thing-counters.jsx]: I decided on:
` const unfiltered = queryKeys.length === 0 || (queryKeys == ['page']);`

The earlier version was overcomplicated, thanks for pointing that out!

File: ui/components/thing-counters.jsx (hunk 10 above)
Comment (tadhg-ohiggins) [inline ui/components/thing-counters.jsx]: Good point, I added that in my last commit.
